src/main/java/com/google/sps/servlets/GoServlet.java

```diff
@@ -0,0 +1,301 @@
+   */
+  public List<String> chooseWaypointCombinationWithShortestTravelTime(
+      String origin, String destination, List<List<String>> allWaypointCombinations)
+      throws DirectionsException {
+    long minTravelTime = 0;
+    List<String> mostOptimalWaypointCombination = new ArrayList<String>();
+    for (List<String> waypointCombination : allWaypointCombinations) {
+      DirectionsClient directionsClient = directionsClientFactory.getDirectionsClient(apiKey);
+      DirectionsResult directionsResult =
+          directionsClient.getDirections(origin, destination, waypointCombination);
+      long travelTime = DirectionsClient.getTotalTravelTime(directionsResult);
+      if (minTravelTime == 0 || travelTime < minTravelTime) {
+        minTravelTime = travelTime;
+        mostOptimalWaypointCombination = waypointCombination;
+      }
+    }
+    return mostOptimalWaypointCombination;
+  }
+
+  /**
+   * Finds the most optimal route of travel between the origin and destination and a set of
+   * waypoints which if the exact location is not known, an exact location is determined and chosen
+   * based on the resulting travel time. Scope of method is public for testing purposes.
+   *
+   * @param origin The starting point of travel.
+   * @param destination The ending point of travel.
+   * @param waypoints The waypoints that should be visited while travelling from the origin to the
+   *     destination.
+   * @return The most optimal set of waypoints between origin and destination.
+   * @throws GeocodingException An exception thrown when an error occurs with the Geocoding API.
+   * @throws PlacesException An exception thrown when an error occurs with the Places API.
+   * @throws DirectionsException An exception thrown when an error occurs with the Directions API.
+   */
+  public List<String> optimizeSearchNearbyWaypoints(
+      String origin, String destination, List<String> waypoints)
+      throws GeocodingException, PlacesException, DirectionsException {
+
+    LatLng originAsCoordinates =
+        GeocodingClient.getCoordinates(
+            geocodingClientFactory.getGeocodingClient(apiKey).getGeocodingResult(origin));
+
+    LatLng destinationAsCoordinates =
+        GeocodingClient.getCoordinates(
+            geocodingClientFactory.getGeocodingClient(apiKey).getGeocodingResult(destination));
+
+    List<String> streetAddressWaypoints = new ArrayList<>();
+    List<LatLng> streetAddressWaypointsAsCoordinates = new ArrayList<>();
+    List<PlaceType> nonStreetAddressWaypointsAsPlaceTypes = new ArrayList<>();
+    separateWaypoints(
+        waypoints,
+        streetAddressWaypoints,
+        streetAddressWaypointsAsCoordinates,
+        nonStreetAddressWaypointsAsPlaceTypes);
+
+    // All street address coordinates including origin and destination are collected
+    List<LatLng> streetAddressesAsCoordinates = streetAddressWaypointsAsCoordinates;
```

Comment:
Hmm it would be misleading to add the origin and destination into a list that is meant for waypoints. I've sort of took that direct assignment out and used `addAll` instead. I don't know if that helps, let me know!

Reply:
Yes, I think this makes it more readable. Thanks!

Reply:
Awesome, thanks!

Reply:
`RESOLVED`